File: scripts/newsletter/detect-image-source.js
#!/usr/bin/env node
// Detect whether an image URL is Substack-hosted and extract its S3 image UUID.
// Usage: node detect-image-source.js "<image-url>"
// Output: JSON { original_url, clean_url, isSubstack, uuid?, innerUrl? }
//
// Substack images are usually served via a CDN wrapper:
//   https://substackcdn.com/image/fetch/$s_!x!,.../https%3A%2F%2Fsubstack-post-media.s3.amazonaws.com%2Fpublic%2Fimages%2F<uuid>_WxH.png
// The publication is NOT encoded in the URL — only the image identity (uuid) is.

const {
  cleanUrl,
  isSubstackImage,
  substackImageUuid,
} = require("./url-utils.js");

const url = process.argv[2];
if (!url) {
  console.error("Usage: node detect-image-source.js <image-url>");
  process.exit(1);
}

// Pull the inner S3 URL out of a substackcdn /image/fetch/ wrapper (if present).
function extractInnerUrl(targetUrl) {
  const marker = targetUrl.indexOf("/https%3A%2F%2F");
  if (marker !== -1) return decodeURIComponent(targetUrl.slice(marker + 1));
  // Some forms embed a plain (already-decoded) inner https URL.
  const plain = targetUrl.indexOf("/https://", 8);
  if (plain !== -1) return targetUrl.slice(plain + 1);
  return targetUrl;
}

const cleaned = cleanUrl(url);
const isSubstack = isSubstackImage(url);
const uuid = isSubstack ? substackImageUuid(url) : null;
const innerUrl = isSubstack ? extractInnerUrl(url) : null;

console.log(JSON.stringify({
  original_url: url,
  clean_url: cleaned,
  isSubstack,
  ...(uuid ? { uuid } : {}),
  ...(innerUrl ? { innerUrl } : {}),
}, null, 2));

File: scripts/newsletter/detect_image_source.go
// Detect whether an image URL is Substack-hosted and extract its S3 image UUID.
// Usage: go run ./scripts/newsletter detect-image-source "<image-url>"
// Output: JSON { original_url, clean_url, isSubstack, uuid?, innerUrl? }
//
// Substack images are usually served via a CDN wrapper:
//
//	https://substackcdn.com/image/fetch/$s_!x!,.../https%3A%2F%2Fsubstack-post-media.s3.amazonaws.com%2Fpublic%2Fimages%2F<uuid>_WxH.png
//
// The publication is NOT encoded in the URL — only the image identity (uuid) is.
package main

import (
	"fmt"
	"net/url"
	"os"
	"strings"
)

// extractInnerURL pulls the inner S3 URL out of a substackcdn /image/fetch/
// wrapper (if present).
func extractInnerURL(target string) string {
	if marker := strings.Index(target, "/https%3A%2F%2F"); marker != -1 {
		decoded, err := url.PathUnescape(target[marker+1:])
		if err != nil {
			return target[marker+1:]
		}
		return decoded
	}
	// Some forms embed a plain (already-decoded) inner https URL.
	if len(target) > 8 {
		if plain := strings.Index(target[8:], "/https://"); plain != -1 {
			return target[8+plain+1:]
		}
	}
	return target
}

type detectImageOutput struct {
	OriginalURL string `json:"original_url"`
	CleanURL    string `json:"clean_url"`
	IsSubstack  bool   `json:"isSubstack"`
	UUID        string `json:"uuid,omitempty"`
	InnerURL    string `json:"innerUrl,omitempty"`
}

func runDetectImageSource(args []string) {
	if len(args) < 1 || args[0] == "" {
		fmt.Fprintln(os.Stderr, "Usage: go run ./scripts/newsletter detect-image-source <image-url>")
		os.Exit(1)
	}
	target := args[0]

	out := detectImageOutput{
		OriginalURL: target,
		CleanURL:    cleanURL(target),
		IsSubstack:  isSubstackImage(target),
	}
	if out.IsSubstack {
		out.UUID = substackImageUUID(target)
		out.InnerURL = extractInnerURL(target)
	}
	printJSON(out)
}
